### ISSUE TITLE
panic in heimdall

### ISSUE THREAD

**Issue by matthewdarwin** [maticnetwork/heimdall#986]
**Heimdall version** (use `heimdall version` or `git rev-parse --verify HEAD` if installed from source):

```
$ heimdalld version
0.3.3
```

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

Linux debian 11

**What happened**:

panic

![image](https://user-images.githubusercontent.com/26171839/234870678-2949ab7c-5e64-429b-91a9-a0e03541d2ae.png)

**What you expected to happen**:

no panic

**Have you tried the latest version**: yes/no

Yes.

**How to reproduce it** (as minimally and precisely as possible):

Not sure if this is reducible or just a one-time thing.  Trying to sync **mumbai** heimdall from scratch

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

```
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.195] Updates to validators                        module=state updates=C275DC8BE39F50D12F66B6A63629C39DA5BAE5BD:6065818
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.206] Committed state                              module=state height=10630663 txs=4 appHash=DC7EB8B5FFD84886AF3D5AFFD8759D83774E892474B2263F54802B14F4B5F5F9
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.211] Executed block                               module=state height=10630664 validTxs=1 invalidTxs=0
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.219] Committed state                              module=state height=10630664 txs=1 appHash=AFA3E336138B65372BF53843C4CA00685286F42889EA7E64C7B7D1A49DBC9952
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.224] Executed block                               module=state height=10630665 validTxs=0 invalidTxs=0
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.234] Committed state                              module=state height=10630665 txs=0 appHash=527A1F1F360476838C5F5490F99229C1BF33AB2247716CBDE93516DC2ABBC74D
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: ERROR[2023-04-26|12:08:29.243] Older invalid tx found                       module=staking
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: ERROR[2023-04-26|12:08:29.244] Older invalid tx found                       module=staking
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: DEBUG[2023-04-26|12:08:29.245] [ENDBLOCK] Updated current validator set     module=app proposer="Validator{2 BE188D6641E8B680743A4815DFA0F6208038960F 0x04888a737a003f4e522ccf23bd9980fdbe7ef2b54365249deba0f9acd45279d66355b1864173b2cf9e75a1cbfb45e65a1a72b9ea76e47aa4bd50d79772ef301769 VP:55410810 A:-61468450}"
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.245] Executed block                               module=state height=10630666 validTxs=3 invalidTxs=2
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.245] Updates to validators                        module=state updates=C275DC8BE39F50D12F66B6A63629C39DA5BAE5BD:6065805
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.251] Committed state                              module=state height=10630666 txs=5 appHash=B8DEF5CE00329225E791DADB817BEC5F0ACDE5BB9C5E25632B6BE890152D3045
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.255] Executed block                               module=state height=10630667 validTxs=0 invalidTxs=0
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.259] Committed state                              module=state height=10630667 txs=0 appHash=B8DEF5CE00329225E791DADB817BEC5F0ACDE5BB9C5E25632B6BE890152D3045
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.264] Executed block                               module=state height=10630668 validTxs=1 invalidTxs=0
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.270] Committed state                              module=state height=10630668 txs=1 appHash=C8FE4576C359A61E0DC6932ED527857C605DDF7A41E979F09BF102FC9D2A30E6
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.275] Executed block                               module=state height=10630669 validTxs=4 invalidTxs=0
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.281] Committed state                              module=state height=10630669 txs=4 appHash=3609D180252ACC5705B83F14C7A1EE83BB63DEDAAD4DCB96FACD014B27F8CA54
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.286] Executed block                               module=state height=10630670 validTxs=1 invalidTxs=1
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: INFO [2023-04-26|12:08:29.291] Committed state                              module=state height=10630670 txs=2 appHash=60C58E8E383EE496FC5430E1344878945D595A5266750C9B2F3FC25598CA5665
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: panic: Failed to process committed block (10630671:06B40BF366D3A029B3A5B2567C2714D46E183427AE35A03E070839B21E225A08): Wrong Block.Header.AppHash.  Expected 60C58E8E383EE496FC5430E1344878945D595A5266750C9B2F3FC25598CA5665, got CCB7437AFC0588D4A9514F52CB2F45FFA08313E489ED380E703FD5C38051F0F9
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: goroutine 178 [running]:
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc00036aa80)
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: #011/home/runner/go/pkg/mod/github.com/maticnetwork/tendermint@v0.26.0-dev0.0.20220923185258-3e7c7f86ce9f/blockchain/v0/reactor.go:344 +0x10c5
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).OnStart
Apr 26 12:08:29 mumbai-arch41 heimdall[23824]: #011/home/runner/go/pkg/mod/github.com/maticnetwork/tendermint@v0.26.0-dev0.0.20220923185258-3e7c7f86ce9f/blockchain/v0/reactor.go:118 +0x7a
Apr 26 12:08:29 mumbai-arch41 systemd[1]: heimdalld.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Apr 26 12:08:29 mumbai-arch41 systemd[1]: heimdalld.service: Failed with result 'exit-code'.
Apr 26 12:08:29 mumbai-arch41 systemd[1]: heimdalld.service: Consumed 4d 2h 17min 35.973s CPU time.
Apr 26 12:08:34 mumbai-arch41 systemd[1]: heimdalld.service: Scheduled restart job, restart counter is at 1.
Apr 26 12:08:34 mumbai-arch41 systemd[1]: Stopped heimdalld.
Apr 26 12:08:34 mumbai-arch41 systemd[1]: heimdalld.service: Consumed 4d 2h 17min 35.973s CPU time.
Apr 26 12:08:34 mumbai-arch41 systemd[1]: Starting heimdalld...
Apr 26 12:08:34 mumbai-arch41 systemd[1]: Started heimdalld.
```

**Config (you can paste only the changes you've made)**:

**node command runtime flags**:

```
$ /usr/bin/heimdalld start --home /var/lib/heimdall
```

**`/dump_consensus_state` output for consensus bugs**

n/a

**Anything else we need to know**:

**Comment by VAIBHAVJINDAL3012** [comment]
@matthewdarwin  Can you please try to run this command ?
/usr/bin/heimdalld rollback --home=/var/lib/heimdall  --chain-id=heimdall-137

**Comment by matthewdarwin** [comment]
```
$ /usr/bin/heimdalld rollback --home=/var/lib/heimdall --chain-id=heimdall-137
Rolled back state to height 10630670 and hash 60C58E8E383EE496FC5430E1344878945D595A5266750C9B2F3FC25598CA5665INFO [2023-04-27|22:51:08.008] Heimdall services stopped                    module=cmd/heimdalld
```

Then restart the service

```
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: INFO [2023-04-27|22:59:00.777] starting ABCI with Tendermint                module=main
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: INFO [2023-04-27|22:59:00.777] Config file /var/lib/heimdall/config/heimdall-config.toml already exists. Skip writing default heimdall config. module=cmd/heimdalld
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: INFO [2023-04-27|22:59:00.778] Found genesis file /var/lib/heimdall/config/genesis.json, skipping writing genesis file
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]:  module=cmd/heimdalld
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: DEBUG[2023-04-27|22:59:00.791] Invalid ETH RPC timeout provided, falling back to default value timeout=5s
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: DEBUG[2023-04-27|22:59:00.791] Invalid BOR RPC timeout provided, falling back to default value timeout=5s
Apr 27 22:59:00 mumbai-arch41 heimdall[249601]: ERROR: failed to create new node: error during handshake: error on replay: Wrong Block.Header.AppHash.  Expected 60C58E8E383EE496FC5430E1344878945D595A5266750C9B2F3FC25598CA5665, got CCB7437AFC0588D4A9514F52CB2F45FFA08313E489ED380E703FD5C38051F0F9
```

**Comment by matthewdarwin** [comment]
@VAIBHAVJINDAL3012 any other suggestions?

**Comment by matthewdarwin** [comment]
Hello Polygon team. Any updates?

**Comment by VAIBHAVJINDAL3012** [comment]
Hi @matthewdarwin  ! Sorry for the delay in response. We are still working to figure out the cause behind this. Will get back to you soon.

**Comment by matthewdarwin** [comment]
I started sync over and got a crash at exactly the same place.

![image](https://github.com/maticnetwork/heimdall/assets/26171839/ce71dce6-fcc6-4913-96cf-6d9cae056eda)

**Comment by VAIBHAVJINDAL3012** [comment]
@matthewdarwin  Can you please share all the configs files you have used?

**Comment by victorfei** [comment]
same issue here:
```
INFO [2023-06-02|21:59:49.740] starting ABCI with Tendermint                module=main
INFO [2023-06-02|21:59:49.740] Config file /user/heimdall/config/heimdall-config.toml already exists. Skip writing default heimdall config. module=cmd/heimdalld
INFO [2023-06-02|21:59:49.741] Found genesis file /user/heimdall/config/genesis.json, skipping writing genesis file
 module=cmd/heimdalld
DEBUG[2023-06-02|21:59:49.749] Invalid ETH RPC timeout provided, falling back to default value timeout=5s
DEBUG[2023-06-02|21:59:49.749] Invalid BOR RPC timeout provided, falling back to default value timeout=5s
ERROR: failed to create new node: error during handshake: error on replay: Wrong Block.Header.AppHash.  Expected B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA, got F614FBD4E1F3AB75D6E288ABC664F4D692AFC591E644829CB0C1AE9DC081E0B8
```

all other solution, rollback, delete data dir doesn't work. Not sure if it has to do with current version. on hemidald 0.33.
I"m not using snapshots.

**Comment by victorfei** [comment]
Follow up: I tried heimdall 0.3.2, still runs into the same issue. I hope the official answer is NOT please download a heimdall snapshot then try to start again; because heimdall should be able to bootstrap from beginning. I used a snapshot before, and heimdall was syncing fine, but I ended up getting missing transactions later in my client when making Json-rpc calls.

**Comment by matthewdarwin** [comment]
@victorfei 100% agree that "download snpshot" is not a real solution.

**Comment by victorfei** [comment]
polygon team any updates?

**Comment by jajaislanina** [comment]
Having the same issue on one of our nodes.

`ERROR: failed to create new node: error during handshake: error on replay: Wrong Block.Header.AppHash.  Expected B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA, got F614FBD4E1F3AB75D6E288ABC664F4D692AFC591E644829CB0C1AE9DC081E0B8`

**Comment by roderik** [comment]
Same here on 0.3.4

```
ERROR: failed to create new node: error during handshake: error on replay: Wrong Block.Header.AppHash.  Expected B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA, got F614FBD4E1F3AB75D6E288ABC664F4D692AFC591E644829CB0C1AE9DC081E0B8
```

**Comment by Raneet10** [comment]
We're digging into this issue on priority. Thanks for your patience and apologies for the inconvenience! Meanwhile, if you wish to urgently fix your nodes, you can always use the snapshots.

**Comment by VAIBHAVJINDAL3012** [comment]
Hi everyone! We have figured out the issue and currently working on the solution of this. Hoping to release the solution by the end of next week.

**Comment by DengLZ** [comment]
**Heimdall version** (use `heimdall version` or `git rev-parse --verify HEAD` if installed from source):

0xpolygon/heimdall   0.3.4

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:
```
> docker images
REPOSITORY             TAG       IMAGE ID             CREATED          SIZE
0xpolygon/bor           0.4.0     dd74e7f4b3e6      2 weeks ago    77.3MB
0xpolygon/heimdall   0.3.4     8a5a084eab4d    4 weeks ago    99.3MB
```

**What happened**:
I downloaded the latest snapshot of heimdall and replaced the data directory in the container with the data directory. Then an error was reported when starting. (The output log still has some garbled characters)
```
/heimdall-home/data/application.db # heimdalld start --home=/heimdall-home
INFO [2023-07-15|14:37:30.887] starting ABCI with Tendermint                module=main 
INFO [2023-07-15|14:37:30.887] Config file /heimdall-home/config/heimdall-config.toml already exists. Skip writing default heimdall config. module=cmd/heimdalld 
INFO [2023-07-15|14:37:30.887] Found genesis file /heimdall-home/config/genesis.json, skipping writing genesis file
 module=cmd/heimdalld 
DEBUG[2023-07-15|14:37:30.910] Invalid ETH RPC timeout provided, falling back to default value timeout=5s
DEBUG[2023-07-15|14:37:30.910] Invalid BOR RPC timeout provided, falling back to default value timeout=5s
panic: Value missing for hash 69cccdfa46f291e319475952b77616779cd30eed5117d53f8873162e9be63f5e corresponding to nodeKey ni���F��GYR�vw���Q�?�s.��?^

goroutine 1 [running]:
github.com/tendermint/iavl.(*nodeDB).GetNode(0xc0013ba3c0, {0xc001923ae0, 0x20, 0x20})
        /go/pkg/mod/github.com/tendermint/iavl@v0.12.4/nodedb.go:78 +0x539
github.com/tendermint/iavl.(*MutableTree).LoadVersion(0xc000d93980, 0xe12f62)
        /go/pkg/mod/github.com/tendermint/iavl@v0.12.4/mutable_tree.go:317 +0x233
github.com/cosmos/cosmos-sdk/store/iavl.LoadStore({0x24d3330?, 0xc000d938c0?}, {0x8?, {0xc000e3b7c0?, 0xc0000e79d0?, 0x24ac1d8?}}, {0xc0000e7a18?, 0xc0018016b0?}, 0x0)
        /go/pkg/mod/github.com/maticnetwork/cosmos-sdk@v0.37.5-0.20220311095845-81690c6a53e7/store/iavl/store.go:33 +0x71
github.com/cosmos/cosmos-sdk/store/rootmulti.(*Store).loadCommitStoreFromParams(0xc000173560, {0x24be750, 0xc000728980}, {0x7f2c25d185b8?, {0xc000e3b7c0?, 0x7f2bfc6aa268?, 0xd0?}}, {{0x24be750, 0xc000728980}, {0x0, ...}, ...})
        /go/pkg/mod/github.com/maticnetwork/cosmos-sdk@v0.37.5-0.20220311095845-81690c6a53e7/store/rootmulti/store.go:447 +0x41e
github.com/cosmos/cosmos-sdk/store/rootmulti.(*Store).LoadVersion(0xc000173560, 0xc000014968?)
        /go/pkg/mod/github.com/maticnetwork/cosmos-sdk@v0.37.5-0.20220311095845-81690c6a53e7/store/rootmulti/store.go:197 +0x305
github.com/cosmos/cosmos-sdk/store/rootmulti.(*Store).LoadLatestVersion(0xc000173560)
        /go/pkg/mod/github.com/maticnetwork/cosmos-sdk@v0.37.5-0.20220311095845-81690c6a53e7/store/rootmulti/store.go:155 +0x35
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).LoadLatestVersion(0xc00077db80, 0xc00070c330?)
        /go/pkg/mod/github.com/maticnetwork/cosmos-sdk@v0.37.5-0.20220311095845-81690c6a53e7/baseapp/baseapp.go:209 +0x2f
github.com/maticnetwork/heimdall/app.NewHeimdallApp({0x24c7cf0, 0xc00020f8c0}, {0x24d3240, 0xc000014968}, {0xc0014136e0, 0x1, 0x1})
        /go/src/github.com/maticnetwork/heimdall/app/app.go:476 +0x683c
github.com/maticnetwork/heimdall/cmd/heimdalld/service.getNewApp.func1({0x24c7cf0, 0xc00020f8c0}, {0x24d3240, 0xc000014968}, {0x1?, 0xc0005baaa0?})
        /go/src/github.com/maticnetwork/heimdall/cmd/heimdalld/service/service.go:219 +0x165
github.com/maticnetwork/heimdall/cmd/heimdalld/service.startInProcess(0x3f?, {0x24c85b0, 0xc000612700}, 0xc000139860, 0xc000fbd940, 0xc000bf6700, 0x0, 0x0)
        /go/src/github.com/maticnetwork/heimdall/cmd/heimdalld/service/service.go:423 +0x334
github.com/maticnetwork/heimdall/cmd/heimdalld/service.heimdallStart.func1(0xc0010a2000?, {0xc000fbde00?, 0x0?, 0x1?})
        /go/src/github.com/maticnetwork/heimdall/cmd/heimdalld/service/service.go:271 +0x16a
github.com/spf13/cobra.(*Command).execute(0xc0010a2000, {0xc000fbdde0, 0x1, 0x1})
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc001069900)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2ef
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/tendermint/tendermint/libs/cli.Executor.Execute({0xc001069900?, 0x1f60fe0?})
        /go/pkg/mod/github.com/maticnetwork/tendermint@v0.26.0-dev0.0.20220923185258-3e7c7f86ce9f/libs/cli/setup.go:89 +0x36
github.com/maticnetwork/heimdall/cmd/heimdalld/service.NewHeimdallService({0x24c6240, 0xc000046038}, {0x0?, 0x0, 0x0})
        /go/src/github.com/maticnetwork/heimdall/cmd/heimdalld/service/service.go:205 +0x6d3
main.main()
        /go/src/github.com/maticnetwork/heimdall/cmd/heimdalld/main.go:10 +0x2e
```
I refer to the following documents. 
https://snapshots.polygon.technology/
https://wiki.polygon.technology/docs/pos/reference/snapshot-instructions-heimdall-bor


**What you expected to happen**:
There may be abnormalities in the snapshot data, or there may be abnormalities in the docker image.

**Have you tried the latest version**: yes/no
yes

**How to reproduce it** (as minimally and precisely as possible):
system: amazone linux2
snapshot: https://snapshots.polygon.technology/
docker image: 0xpolygon/heimdall   0.3.4

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:
```
INFO [2023-07-15|14:37:30.887] Found genesis file /heimdall-home/config/genesis.json, skipping writing genesis file
 module=cmd/heimdalld 
DEBUG[2023-07-15|14:37:30.910] Invalid ETH RPC timeout provided, falling back to default value timeout=5s
DEBUG[2023-07-15|14:37:30.910] Invalid BOR RPC timeout provided, falling back to default value timeout=5s
panic: Value missing for hash 69cccdfa46f291e319475952b77616779cd30eed5117d53f8873162e9be63f5e corresponding to nodeKey ni���F��GYR�vw���Q�?�s.��?^
goroutine 1 [running]:
github.com/tendermint/iavl.(*nodeDB).GetNode(0xc0013ba3c0, {0xc001923ae0, 0x20, 0x20})
        /go/pkg/mod/github.com/tendermint/iavl@v0.12.4/nodedb.go:78 +0x539
github.com/tendermint/iavl.(*MutableTree).LoadVersion(0xc000d93980, 0xe12f62)
```

**Config (you can paste only the changes you've made)**:
Refer to the official website
**node command runtime flags**:
Refer to the official website
**`/dump_consensus_state` output for consensus bugs**
panic: Value missing for hash 69cccdfa46f291e319475952b77616779cd30eed5117d53f8873162e9be63f5e corresponding to nodeKey ni���F��GYR�vw���Q�?�s.��?^

**Anything else we need to know**:
Without using snapshot data, the service can normally synchronize data.

**Comment by Raneet10** [comment]
Hey @DengLZ , can you open a separate issue for this ? Seems unrelated to the issue being discussed here and it'll also help us to track better.

**Comment by skolem-dan** [comment]
I've been seeing this same issue on a couple nodes I've been trying to sync for the past few weeks. Are there any updates to share?

It sounds like the team may already have a lead on the cause, but just in case here's a bit more detail that I haven't seen others share:

```
INFO [2023-07-16|05:08:41.742] Committed state                              module=state height=8587010 txs=0 appHash=B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA
INFO [2023-07-16|05:08:41.752] Executed block                               module=state height=8587011 validTxs=0 invalidTxs=0
INFO [2023-07-16|05:08:41.762] Committed state                              module=state height=8587011 txs=0 appHash=B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA
INFO [2023-07-16|05:08:41.772] Executed block                               module=state height=8587012 validTxs=0 invalidTxs=1
INFO [2023-07-16|05:08:41.782] Committed state                              module=state height=8587012 txs=1 appHash=B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA
panic: Failed to process committed block (8587013:3D5D7EEECE3143EFAA92A59B295738FF69D0FA41433FA6892F23F0B2CB96C7FB): Wrong Block.Header.AppHash.  Expected B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA, got F614FBD4E1F3AB75D6E288ABC664F4D692AFC591E644829CB0C1AE9DC081E0B8

goroutine 111 [running]:
github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc000ded500)
        /go/pkg/mod/github.com/maticnetwork/tendermint@v0.26.0-dev0.0.20220923185258-3e7c7f86ce9f/blockchain/v0/reactor.go:344 +0x10c5
created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).OnStart
        /go/pkg/mod/github.com/maticnetwork/tendermint@v0.26.0-dev0.0.20220923185258-3e7c7f86ce9f/blockchain/v0/reactor.go:118 +0x7a
INFO [2023-07-16|05:08:42.282] starting ABCI with Tendermint                module=main
INFO [2023-07-16|05:08:42.282] Config file /data/heimdall/config/heimdall-config.toml already exists. Skip writing default heimdall config. module=cmd/heimdalld
INFO [2023-07-16|05:08:42.282] Found genesis file /data/heimdall/config/genesis.json, skipping writing genesis file
 module=cmd/heimdalld
DEBUG[2023-07-16|05:08:42.528] Invalid ETH RPC timeout provided, falling back to default value timeout=5s
DEBUG[2023-07-16|05:08:42.528] Invalid BOR RPC timeout provided, falling back to default value timeout=5s
ERROR: failed to create new node: error during handshake: error on replay: Wrong Block.Header.AppHash.  Expected B524A829BFCB8ABB78D0BEB796DF5E770C4C23DDB632F6B6CB6048936F9161CA, got F614FBD4E1F3AB75D6E288ABC664F4D692AFC591E644829CB0C1AE9DC081E0B8
```

This is happening on servers using the same config that's worked on others that were synced earlier this year and are still working fine, so I'm reasonably certain it's not a configuration issue. I've also tried:

- Stopping Heimdall, `heimdalld unsafe-reset-all`, downloading and extracting fresh snapshots, restarting Heimdall
- Stopping Heimdall, `heimdalld unsafe-reset-all`, restarting Heimdall to sync from scratch
- Stopping Heimdall, wiping the drive, restoring just the basic config, downloading and extracting fresh snapshots, restarting Heimdall

All of these methods have resulted in the same behavior, and always get stuck at the same block.

**Comment by VAIBHAVJINDAL3012** [comment]
@skolem-dan  We have figured out the issue and implemented the solution for it. We are going to release it in the first part of next week. But if you want to sync at urgent, use this branch vaibhav/HeimdallSyncIssue

**Comment by matthewdarwin** [comment]
@VAIBHAVJINDAL3012 any news?

**Comment by VAIBHAVJINDAL3012** [comment]
@matthewdarwin We have solved this issue. As for now you can use this e7a630251796c91494d6cec2ef2eafe8799cff42 branch hash. We will release the tag soon.
Thank you for bringing this issue to our notice.

**Comment by VAIBHAVJINDAL3012** [comment]
@skolem-dan Can you please send you flag or service file?

**Comment by skolem-dan** [comment]
I'm using the `0xpolygon/heimdall:0.3.5-beta-1` Docker image, and starting Heimdall with `/usr/bin/heimdalld start --home=/data/heimdall`

I'm using docker-compose, so here's the relevant portion of that file:

```
  heimdall:
    image: 0xpolygon/heimdall:0.3.5-beta-1
    entrypoint: /usr/bin/heimdalld
    command: start --home=/data/heimdall
    user: 1000:1000
    cap_drop:
      - ALL
    restart: always
    tmpfs:
      - /tmp
    volumes:
      - /data/heimdall:/data/heimdall
    ports:
      - 127.0.0.1:26656:26656
      - 127.0.0.1:26657:26657
```

`/data/heimdall` looks pretty normal to me at a glance:

```
$ docker-compose exec heimdall ls -l /data/heimdall
total 8960
drwxr-xr-x    2 1000     wheel       405504 Aug 14 18:06 application.db
drwxr-xr-x    2 1000     wheel      5894144 Aug 14 18:06 blockstore.db
drwxrwx---    2 1000     wheel          194 Aug 16 06:42 config
drwx------    2 1000     wheel        53248 Aug 14 18:06 cs.wal
drwxrwx---    7 1000     wheel          164 Aug 14 18:06 data
drwxr-xr-x    2 1000     wheel         4096 Aug 14 18:06 evidence.db
-rw-------    1 1000     wheel           48 Oct  5  2020 priv_validator_state.json
drwxr-xr-x    2 1000     wheel       221184 Aug 14 18:06 state.db
drwxr-xr-x    2 1000     wheel        57344 Aug 14 18:06 tx_index.db
$ docker-compose exec -u root heimdall du -sh /data/heimdall
566G	/data/heimdall
```

**Comment by VAIBHAVJINDAL3012** [comment]
Can you please use this command-
command: start --home=/data/heimdall --chain=mainnet

**Comment by VAIBHAVJINDAL3012** [comment]
Actually you need to restart it from the snapshot.

**Comment by skolem-dan** [comment]
Alright, I can try that and provide an update in a couple days when it reaches that block again.

Is this a newly required flag? I've successfully synced Heimdall in the past without it, using the same docker-compose file I provided above.

**Comment by VAIBHAVJINDAL3012** [comment]
Yes, It is a required flag

**Comment by skolem-dan** [comment]
Quick update: Adding `--chain=mainnet` seems to have gotten things working again:

```
$ curl -s http://127.0.0.1:26657/status | jq .result.sync_info
{
  "latest_block_hash": "69DDA621E018BEDDA1122A376D2DB25156EAC54E1E005A18F326D181651C8ED3",
  "latest_app_hash": "373627CF9CFE34BE271B2FA97BD9EC7F6B8FC203FFB03B553746D46D7B9EDA89",
  "latest_block_height": "15264600",
  "latest_block_time": "2023-08-21T22:00:43.280298207Z",
  "catching_up": false
}
```

I'm still waiting on erigon snapshots to be available again before I can finish syncing these nodes, but it does look like at least Heimdall is working again. Thanks for the help!
